### PR TITLE
Documentation Parser Extended Integration 

### PR DESCRIPTION
**Pull request by BinarySoftware**
## TO DO LIST
 - [x] Parsing code containing comment with function defined in next line -
   nothing more complicated. In result we want documentation just like we get
    now, but with title created from function name. After updating code to
    newest parser ( lets assume wdanilo/cata branch is your reference now on)
    make a PR onto it. It may be dirty here and there, but it should work

-  [x] Add to our AST node "documented" which will contain two fields - other AST
    and your Documentation. Your doc shouldn't extend AST ( it doesn't on my
    branch already ) - separate PR 
   ( RE: Cannot be in separate PR as it is vital for functionality of Doc Parser creator )

-  [x] Now use case working like that - in code we have `def Maybe` with doc
    above it and to every function in it (refer to parser tests). We want to
    traverse it and make final form of documentation. Take from def name of
    documentation, show documentation of every internal function just like
    apple displays on grey background in Developer.Apple.com Documentation.
    Generally implementation here should be extremely easy as def is block
    which contains list of lines so you loop through it and if you find Doc
    then check if next line is empty or contain function ( and you have it
    already from point 1 )

#### For reference - (#121)
This is really important, as it deeply integrates together DocParser with Parser. Now if you have some code like that:
```
## This function adds together x and y
foo x y = x + y
```
it will output to you:
```
Documentation(Documented(This function adds together x and y),AST(...))
```
In short words - it combines together parser output with docParser output in a better way, creates titles for documentation out of function names.

Also I've upgraded HTML generator with CSS to reflect changes, here's example output:
Zrzut ekranu 2019-08-30 o 17 46 47

### Important Notes
- As it is a small PR it can only create title from function name, some methods may be generalised in future

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 1]
```diff
@@ -49,3 +75,177 @@ object DocParser {
     writer.close()
   }
 }
+
+// BROKEN AFTER MERGE OF AST_CATA BRANCH - PARSER WONT PRODUCE COMMENTS!
```
Comment (iamrecursion) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Just a note, even if they are `TODO` comments for yourself, pleae keep them in english.

Comment (BinarySoftware) [reply syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Ok, I've just copy-pasted what Wojtek sent to me yesterday, It is just a small note for me, but I can translate it

Comment (BinarySoftware) [reply syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Ok, I've just fixed that @iamrecursion

Comment (iamrecursion) [reply syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Brilliant, thanks!

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 2]
```diff
@@ -8,35 +8,59 @@ import org.enso.flexer.Reader
...
-      saveHTMLCodeToLocalFile(path, v.renderHTML("style.css"))
-      v
-    case _ => Doc()
+  def runMatched(input: String): Doc = run(input) match {
+    case flexer.Parser.Result(_, flexer.Parser.Result.Success(v)) => v
+    case _                                                        => Doc()
   }
   def run(input: String): Result[Doc] = engine.run(new Reader(input))
 
+  def onHTMLRendering(ast: AST, doc: Doc): Unit = {
```
Comment (iamrecursion) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: As should all the functions.

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 3]
```diff
@@ -49,3 +73,232 @@ object DocParser {
...
+    */
+  def swapInfixWithDocumentedIntoDocs(
+    currIndex: Int,
+    documented: AST.DocumentedOf[AST],
+    astDoc: List1[AST.Block.OptLineOf[AST]],
+    astFromParser: List1[AST.Block.OptLineOf[AST]]
+  ): Option[List1[AST.Block.OptLineOf[AST]]] = {
+    documented.ast.unFix match {
+      case AST.Ident.ConsOf("") => None
+      // NOTE : Documented without any AST -> No infix after -> nothing to do
```
Comment (iamrecursion) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: These notes should be translated to use the [source notes](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md#source-notes) format and expanded upon a bit. See [here](https://github.com/luna/enso/blob/master/Interpreter/src/main/java/org/enso/interpreter/node/callable/argument/ReadArgumentNode.java#L50) for an example of how they work in practice.

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 4]
```diff
@@ -49,3 +73,232 @@ object DocParser {
...
+        val updatedWithDoc = astDoc.toList.updated(prevIndex, DocToLine)
+        // NOTE : 2nd Documented ( with title ) -> Infix (to get back func. def)
+        val infix            = astFromParser.toList(currIndex)
+        val updatedWithInfix = updatedWithDoc.updated(currIndex, infix)
+        List1(updatedWithInfix)
+    }
+  }
+
+  /** generateHTMLForEveryDocumentation
+    * this method is used for generation of
```
Comment (iamrecursion) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Why is this comment wrapped so strangely? The method name shouldn't be part of the doc comment as it gets out of sync (as this proves).

File: syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala [hunk 5]
```diff
@@ -17,7 +17,7 @@ import org.scalatest.Assertion
 //    output match {
```
Comment (iamrecursion) [inline syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala]: Why are all the tests commented out?

Comment (BinarySoftware) [reply syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala]: They haven't been commented out by me, I think that they were by Wojtek just for ease of reading through tests of new Parser with catamorphic AST

File: syntax/definition/src/main/scala/org/enso/syntax/text/AST2.scala [hunk 6]
```diff
@@ -1181,6 +1181,38 @@ object AST {
...
+  //////////////////////////////////////////////////////////////////////////////
+  //// Documented //////////////////////////////////////////////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+
+  type Documented = ASTOf[DocumentedOf]
+  /*
+    FIXME [MM] : This doesn't work - `t.ast` - with :
+                 case class DocumentedOf[T](ast: T, doc: Doc) extends ShapeOf[T]
+                 Hence replaced `T` with `AST`
+   */
```
Comment (wdanilo) [inline syntax/definition/src/main/scala/org/enso/syntax/text/AST2.scala]: What do you mean it does not work? What happens if you use `T` instead of hardcoded `AST` there? We need to use `T` there.

File: syntax/definition/src/main/scala/org/enso/syntax/text/AST2.scala [hunk 7]
```diff
@@ -1181,6 +1181,38 @@ object AST {
...
+  }
+
+  //// Instances ////
+
+  object DocumentedOf {
+    import Documented._
+    implicit def functor[T]: Functor[DocumentedOf] = semi.functor
+    implicit def repr[T]: Repr[DocumentedOf[T]] = t => {
+      val tRepr: Repr.Builder = {
+        t.ast.show().length match {
```
Comment (wdanilo) [inline syntax/definition/src/main/scala/org/enso/syntax/text/AST2.scala]: `t.ast.show().length` is slow, we should rather not use it this way. Why do you check it for length? Could you please provide me with an example case when the length of show would be 0 here?

File: syntax/definition/src/main/scala/org/enso/syntax/text/AST2.scala [hunk 8]
```diff
@@ -1181,6 +1181,38 @@ object AST {
...
+          case 0 => R
+          case _ => R + " " + t.ast + newline
+        }
+      }
+      R + symbol + symbol + tRepr + t.doc
+    }
+    implicit def offsetZip[T]: OffsetZip[DocumentedOf, T] = _.map((0, _))
```
Comment (wdanilo) [inline syntax/definition/src/main/scala/org/enso/syntax/text/AST2.scala]: Are you sure this is correct? Is the `doc` before `ast` or after `ast`? I mean, this implementation means that `AST` offset from the beginning of `DocumentedOf` block is always `0`, so it means that `doc` is after `ast`, which seems incorrect. Moreover, if Im right, please reschuffle the fields in the DocumentedOf definition so they are in order as well :)

File: syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala [hunk 9]
```diff
@@ -3,30 +3,49 @@ package org.enso.syntax.text.ast
...
   *
-  * Doc can be made of up to 3 elements:
-  * @param tags - If exists, holds applied tags to documented text
-  * @param synopsis - If exists, holds synopsis of documented text
-  * @param body - If exists, holds body of documented text
+  * Documentation can be made of up to 3 elements:
+  *
+  * @param tags     - If exists, holds [[Doc#Tags]] to documented text
+  * @param synopsis - If exists, holds [[Doc#Synopsis]] of documented text
+  * @param body     - If exists, holds [[Doc#Body]] of documented text
```
Comment (wdanilo) [inline syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala]: ❤️

File: syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala [hunk 10]
```diff
@@ -100,11 +88,11 @@ object Doc {
...
 
-    /* Text - used to hold normal string as Elem
-     * Newline - used to hold newline ('\n') as elem
-     */
+    /** Text - used to hold normal string as Elem
+      * Newline - used to hold newline ('\n') as elem
+      */
     final case class Text(text: String) extends Elem {
-      val repr: Repr.Builder = text
+      val repr: Repr.Builder = R + text
```
Comment (wdanilo) [inline syntax/definition/src/main/scala/org/enso/syntax/text/ast/Doc.scala]: why this change? Previous implementation was shiorter and doing the same thing.

File: syntax/definition/src/main/scala/org/enso/syntax/text/spec/DocParserDef.scala [hunk 11]
```diff
@@ -524,8 +524,8 @@ case class DocParserDef() extends Parser[Doc] {
 
   val NEWLINE: State = state.define("Newline")
 
-  ROOT    || newline              || reify { state.begin(NEWLINE) }
-  NEWLINE || indent.EOFPattern    || reify { indent.onEOFPattern() }
+  ROOT    || newline              || reify { state.begin(NEWLINE)     }
+  NEWLINE || indent.EOFPattern    || reify { indent.onEOFPattern()    }
   NEWLINE || indent.indentPattern || reify { indent.onIndentPattern() }
```
Comment (wdanilo) [inline syntax/definition/src/main/scala/org/enso/syntax/text/spec/DocParserDef.scala]: Is this formatting passing scalafmt?

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 12]
```diff
@@ -8,44 +8,400 @@ import org.enso.flexer.Reader
...
+    * function for invoking DocParser in right places
+    * creating Docs from parsed comments
+    * and also generating HTML files for created Doc's
+    *
+    * @param ast - parsed data by Parser
+    * @return - AST with possible documentation
+    */
+  def create(ast: AST): AST = {
+    val createdDocs = createDocs(ast)
+    val preparedDocs = createdDocs.unFix match {
```
Comment (wdanilo) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Dont use `unFix` please. This is internal implementation and it may change any time. Use `case AST.Module.any(m) => ...` instead.

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 13]
```diff
@@ -8,44 +8,400 @@ import org.enso.flexer.Reader
...
+      case AST.ModuleOf(m) =>
+        ast.unFix match {
+          case AST.ModuleOf(v) => reorganiseDocs(m, v)
```
Comment (wdanilo) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: `reorganiseDocs` is not the best name we can have here

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 14]
```diff
@@ -8,44 +8,400 @@ import org.enso.flexer.Reader
...
+    * Doc(s), and Infix into Documented(s) title
+    *
+    * @param ast - data from Parser
+    * @return - modified data containing possibly Documentation(s)
+    */
+  def createDocs(ast: AST): AST = {
+    ast.map { elem =>
+      previousElement = Some(elem.unFix match {
+        case AST.CommentOf(v)        => createDocFromComment(AST.Comment(v))
+        case v: AST.App.InfixOf[AST] => infixFoundWhileCreatingDocs(v)
```
Comment (wdanilo) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: The same applies here - change it to `case AST.App.Infix.any(v) => ...` and dont use `unFix`

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 15]
```diff
@@ -8,44 +8,400 @@ import org.enso.flexer.Reader
...
+  /**
+    * This function changes single- and multi- line comments into
+    * Doc(s), and Infix into Documented(s) title
+    *
+    * @param ast - data from Parser
+    * @return - modified data containing possibly Documentation(s)
+    */
+  def createDocs(ast: AST): AST = {
+    ast.map { elem =>
+      previousElement = Some(elem.unFix match {
```
Comment (wdanilo) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Dont use so long expressions. If you want to pack result in `Some` do it in a separate var

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 16]
```diff
@@ -8,44 +8,400 @@ import org.enso.flexer.Reader
...
+        case AST.CommentOf(v)        => createDocFromComment(AST.Comment(v))
+        case v: AST.App.InfixOf[AST] => infixFoundWhileCreatingDocs(v)
+//        case v: AST.DefOf[AST] =>
+//          defFoundWhileCreatingDocs(v)
+//        // Maybe just call create(v) ?
+        case _ => createDocs(elem)
+      })
+      previousElement.get
```
Comment (wdanilo) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: This is nontotal function. Banned.

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala (hunk 15 above)
Comment (wdanilo) [reply syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Btw, its really hard to understand the code when the logic is written this way. Its asssigning previousElement, while checking current element and when current element is infxi it uses `previousElement` (but not this one, but previous value) to do something. Ugh. Divide it to separate variables and lines :)

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 17]
```diff
@@ -8,44 +8,400 @@ import org.enso.flexer.Reader
...
+    * creates Docs from comments found in parsed data
+    *
+    * @param ast - comment
+    * @return - Documentation
+    */
+  def createDocFromComment(
+    ast: AST.CommentOf[AST]
+  ): AST.DocumentedOf[AST] = {
+    val in = ast.lines.mkString("\n")
+    // FIXME [MM] : (AST.Cons("")) HOW DO I CREATE EMPTY AST()?
```
Comment (wdanilo) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: There is no "empty AST"

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 18]
```diff
@@ -8,44 +8,400 @@ import org.enso.flexer.Reader
...
+    AST.Documented(AST.Cons(""), DocParser.runMatched(in))
+  }
+
+  /**
+    * Tries to create Doc Title from function name
+    *
+    * @param ast - Infix
+    * @return - Documentation title from infix left argument
+    */
+  def createDocumentedWithTitleFromInfix(
```
Comment (wdanilo) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: `createDocumentedWithTitleFromInfix` - its almost perfect. But not semantically corect. You are not creating title from infix, but from value definition right? Value definition is encoded as infox in ASt.

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 19]
```diff
@@ -8,44 +8,400 @@ import org.enso.flexer.Reader
...
+    ast: AST.App.InfixOf[AST],
+    partialDoc: AST.DocumentedOf[AST]
+  ): AST.DocumentedOf[AST] = {
+    AST.Documented(ast.larg, partialDoc.doc)
```
Comment (wdanilo) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: I dont get what it does. It takes infix like `a + b` , throws away everything but `a` and assigns it with a doc like `Documented("a", someDoc)`. When creating `AST.Documented` you should never modify AST, you should only build doc.

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 20]
```diff
@@ -8,44 +8,400 @@ import org.enso.flexer.Reader
...
+    * Tries to create Doc Title from def function name
+    *
+    * @param ast - Def
+    * @return - Documentation title from def name
+    */
+  def createDocumentedWithTitleFromDef(
+    ast: AST.DefOf[AST],
+    partialDoc: AST.DocumentedOf[AST]
+  ): AST.DocumentedOf[AST] = {
+    AST.Documented(ast.name, partialDoc.doc)
```
Comment (wdanilo) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: As with the previosu function, don't modify AST. Modify Doc only. The only operation that you should do with AST is adding `Documented` here and there. Here you are taking `ast` apart.

File: syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 21]
```diff
@@ -8,44 +8,400 @@ import org.enso.flexer.Reader
...
+  }
+
+  /**
+    * reorganizes documented AST with properly placed [[AST.Documented]]
```
Comment (wdanilo) [inline syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: So what does it exactly do? Why do we have to "reorganize AST"? AST was organized after parser passes, then you traverse it and build doc, why do you need to "reorganize" it ?

File: syntax/definition/src/main/scala/org/enso/syntax/text/AST2.scala (hunk 7 above)
Comment (BinarySoftware) [reply syntax/definition/src/main/scala/org/enso/syntax/text/AST2.scala]: 0 - when there is no title/ast, then there is a placeholder - AST.Cons(""), and after that we don't want newline

File: syntax/definition/src/main/scala/org/enso/syntax/text/spec/DocParserDef.scala (hunk 11 above)
Comment (BinarySoftware) [reply syntax/definition/src/main/scala/org/enso/syntax/text/spec/DocParserDef.scala]: yup

**Comment by BinarySoftware** [comment]
I'm completely rewriting it, hence resolving most of comments

**Comment by BinarySoftware** [comment]
I should close this PR and open a new one onto parser-ast-cata-merge branch
